docs/doxygen/man/man3/GetVipKeys.3
.TH "GetVipKeys" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetVipKeys \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetVipKeys\fP (uint16_t abank, uint32_t auser, uint32_t time, uint8_t vhash[32])"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_VIP as type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "unsigned char * \fBgetVipHash\fP ()"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 9 of file getvipkeys\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetVipKeys::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 61 of file getvipkeys\&.cpp\&.
.SS "bool GetVipKeys::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 89 of file getvipkeys\&.cpp\&.
.SS "void GetVipKeys::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 57 of file getvipkeys\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getvipkeys.cpp
#include "getvipkeys.h"
#include <sys/stat.h>
#include "helper/hash.h"
#include "helper/json.h"
#include "hash.hpp"

GetVipKeys::GetVipKeys()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

GetVipKeys::GetVipKeys(uint16_t abank, uint32_t auser, uint32_t ttime, uint8_t vhash[32])
    : m_data{abank, auser, ttime, vhash} {
    m_responseError = ErrorCodes::Code::eNone;
}

int GetVipKeys::getType() {
    return TXSTYPE_VIP;
}

CommandType GetVipKeys::getCommandType() {
    return CommandType::eReadingOnly;
}

unsigned char* GetVipKeys::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetVipKeys::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void GetVipKeys::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetVipKeys::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int GetVipKeys::getDataSize() {
    return sizeof(m_data.info);
}

int GetVipKeys::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* GetVipKeys::getSignature() {
    return m_data.sign;
}

int GetVipKeys::getSignatureSize() {
    return sizeof(m_data.sign);
}

void GetVipKeys::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetVipKeys::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

uint32_t GetVipKeys::getTime() {
    return m_data.info.ttime;
}

uint32_t GetVipKeys::getUserId() {
    return m_data.info.auser;
}

uint32_t GetVipKeys::getBankId() {
    return m_data.info.abank;
}

int64_t GetVipKeys::getFee() {
    return 0;
}

int64_t GetVipKeys::getDeduct() {
    return 0;
}

uint32_t GetVipKeys::getUserMessageId() {
    return 0;
}

bool GetVipKeys::send(INetworkClient& netClient)
{
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetVipKeys sending error\n");
        return false;
    }

    if(!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetVipKeys reading error\n");
        return false;
    }

    if(m_responseError) {
        return true;
    }

    uint32_t bufferLength;

    if(!netClient.readData((int32_t*)&bufferLength, sizeof(bufferLength))) {
        ELOG("GetVipKeys ERROR reading response buffer length\n");
        return false;
    }

    auto responseBuffer = std::make_unique<char[]>(bufferLength);
    if(!netClient.readData(responseBuffer.get(), bufferLength)) {
        ELOG("GetVipKeys ERROR reading response buffer\n");
        return false;
    }

    m_vipKeys.loadFromBuffer(std::move(responseBuffer), bufferLength);

    char hash[65];
    hash[64]='\0';
    ed25519_key2text(hash, getVipHash(), 32);

    if(!m_vipKeys.checkVipKeys(getVipHash(), bufferLength/(2+32))) {
        fprintf(stderr, "ERROR, failed to check VIP keys for hash %s\n", hash);
        return false;
    }

    mkdir("vip", 0755);
    char filename[128];
    sprintf(filename, "vip/%64s.vip", hash);

    if(!m_vipKeys.storeVipKeys(filename)) {
        fprintf(stderr, "ERROR, failed to save VIP keys for hash %s\n", hash);
        return false;
    }

    return true;
}

void GetVipKeys::saveResponse(settings& /*sts*/) {
}

std::string GetVipKeys::toString(bool /*pretty*/) {
    return "";
}

void GetVipKeys::toJson(boost::property_tree::ptree &ptree) {
    char hash[65];
    hash[64]='\0';
    ed25519_key2text(hash, getVipHash(), 32);

    if (!m_responseError) {
        ptree.put("viphash",hash);
        boost::property_tree::ptree viptree;
        for(char* p=m_vipKeys.getVipKeys(); p<m_vipKeys.getVipKeys()+m_vipKeys.getLength(); p+=32+sizeof(uint16_t)) {
            uint16_t srvid = (uint16_t)*p;
            ed25519_key2text(hash, (uint8_t*)(p+2), 32);
            boost::property_tree::ptree vipkey;
            vipkey.put("server_id", srvid);
            vipkey.put("public_key", hash);
            viptree.push_back(std::make_pair("", vipkey));
        }
        ptree.add_child("vipkeys", viptree);
    } else {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void GetVipKeys::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.abank);
    ptree.put(TAG::SRC_USER, m_data.info.auser);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::HASH, ed25519_key2text(m_data.info.viphash, 32));
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

unsigned char* GetVipKeys::getVipHash() {
    return reinterpret_cast<unsigned char*>(&m_data.info.viphash);
}

std::string GetVipKeys::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_vipkeys\",\"viphash\":<viphash 32-byte hex string>}" << "\n";
    ss << "Example: " << "{\"run\":\"get_vipkeys\",\"viphash\":\"D3FD529F6305F574BA22F3BDF761B4778094CB38958300ACDF21D35BE03BDC4F\"}" << "\n";
    return ss.str();
}
